### PR TITLE
[TVF-465] Update WY form filling to work with new form

### PR DESCRIPTION
**Pull request by dzcode**
https://democracyworks.atlassian.net/browse/TVF-465

The fields  in the ticket are not correct - the fields we can use are:

```
 "apt"
 "change_name"
 "change_registration"
 "city"
 "dob_day"
 "dob_month"
 "dob_year"
 "drivers_license"
 "email"
 "first"
 "last"
 "mailing_city"
 "mailing_state"
 "mailing_street_apt"
 "mailing_zip"
 "middle"
 "new_application"
 "old_city"
 "old_county"
 "old_full_name"
 "old_party"
 "old_state"
 "old_street_apt"
 "old_zip"
 "party_constitution"
 "party_democratic"
 "party_libertarian"
 "party_none"
 "party_republican"
 "phone_area_code"
 "phone_first_three"
 "phone_last_four"
 "street"
 "suffix"
 "wy_old_county"
 "wy_old_full_address"
 "wy_old_full_mailing_address"
 "zip"
```

### CODE REVIEW

File: src/voter_registration_works/reg_form/wy.clj [hunk 1]
```diff
@@ -4,6 +4,19 @@
             [turbovote.resource-config :refer [config]]
             [com.climate.newrelic.trace :refer [defn-traced]]))
 
+(defn phone->ctx [phone]
+  (when (and phone (re-matches #"\+?1?\d{10}" phone))
```
Comment (mrichards42) [inline src/voter_registration_works/reg_form/wy.clj:8]: you'd want to capture `\d{10}`, otherwise numbers with `+1` will include that prefix, which you don't want

File: src/voter_registration_works/reg_form/utils.clj [hunk 2]
```diff
@@ -43,3 +45,13 @@
               java.lang.Number [ky (str val)]
               [ky val])))
         identity mp))
+
+(def UTC (ZoneId/of "UTC"))
+
+(defn inst->local-date
+  "Converts a Date (in UTC) to a LocalDate."
+  [date]
```
Comment (mrichards42) [inline src/voter_registration_works/reg_form/utils.clj:53]: Tiny thing -- when you know the java type of your arg, you can add it as an annotation, like `[^Date date]` . . . you don't usually need to do that, but IMO for date stuff it can be helpful documentation. You do get a tiny perf boost since you don't end up using reflection to call methods, but that's really minor unless you're in the middle of a hot path.

**Comment by dzcode** [comment]
This is still a WIP post-changes -- going to test locally and then re-request review

File: resources/config.edn [hunk 3]
```diff
@@ -99,4 +99,4 @@
             :election-authority-cache-ttl 300000}
  :templates {:nvrf {:en "https://s3-us-west-2.amazonaws.com/voter-registration-works/nvrf.en.pdf"
                     :es "https://s3-us-west-2.amazonaws.com/voter-registration-works/nvrf.es.pdf"}
-             :wy {:en "https://s3-us-west-2.amazonaws.com/voter-registration-works/wy.en.pdf"}}}
+             :wy {:en "https://s3-us-west-2.amazonaws.com/voter-registration-works/wy-staging.en.pdf"}}}
```
Comment (dzcode) [inline resources/config.edn:102]: This will be removed before merging.

File: src/voter_registration_works/reg_form/utils.clj [hunk 4]
```diff
@@ -2,15 +2,53 @@
...
-            [voter-registration-works.channels :as channels]
-            [voter-registration-works.states :as states]
-            [voter-registration-works.util :as util]
             [turbovote.resource-config :refer [config]]
-            [com.climate.newrelic.trace :refer [defn-traced]]))
+            [com.climate.newrelic.trace :refer [defn-traced]])
+  (:import (java.util Date)
+           (java.time ZoneId)))
+
+(defn age-at
```
Comment (dzcode) [inline src/voter_registration_works/reg_form/utils.clj:11]: These functions are just moved, but otherwise unchanged.

File: src/voter_registration_works/reg_form/wy.clj [hunk 5]
```diff
@@ -1,33 +1,67 @@
...
+      :city (:city register-address)
+      :zip (some #(% register-address) [:postal-code :zip])}
+     (when (< 8 (count apt))
+       {:street (str street ", " apt)
+        :apt nil}))))
 
 (defn mailing-address->ctx [mailing-address]
-  {:mailing_street_apt (clean-join (:street mailing-address)
-                                   (:street-2 mailing-address))
+  {:mailing_street_apt (clean-join (:street mailing-address) (:street-2 mailing-address))
```
Comment (mrichards42) [inline src/voter_registration_works/reg_form/wy.clj:64]: Nit: looks like an unnecessary whitespace change? (and makes for a pretty long line)

File: src/voter_registration_works/reg_form/wy.clj [hunk 6]
```diff
@@ -66,24 +104,27 @@
      :context context}))
 
 (defn-traced user->ctx
```
Comment (mrichards42) [inline src/voter_registration_works/reg_form/wy.clj:106]: if we're adding `election`, might be good to call this `user-and-election->ctx`

File: src/voter_registration_works/reg_form/wy.clj [hunk 7]
```diff
@@ -1,33 +1,67 @@
...
+
+(defn user->citizenship-ctx [user]
+  {:citizen_yes (when (true? (:us-citizen user)) "On")
+   :citizen_no (when (false? (:us-citizen user)) "On")})
+
+(defn voter->citizenship-ctx [voter]
+  ;; If we know a user's country of citizenship, fill these out
+  (when-let [citizen-country (:citizenship voter)]
+    {:citizen_yes (when (= "US" citizen-country) "On")
+     :citizen_no (when (not= "US" citizen-country) "On")}))
```
Comment (mrichards42) [inline src/voter_registration_works/reg_form/wy.clj:21]: if we have a user w/o a value for this field, I think we would want to leave it blank, like we do in `user->citizenship-ctx`

Comment (dzcode) [reply src/voter_registration_works/reg_form/wy.clj:21]: Wouldn't it be nil in that case because of the `when-let`?

File: src/voter_registration_works/reg_form.clj [hunk 8]
```diff
@@ -135,7 +135,7 @@
                           (:district-division electorate))
                          (user-registration-state user))]
     (condp = form-type
-      :wy (wy/user-and-election->reg-form-ctx user (:election electorate))
+      :wy-reg (wy/user-and-election->reg-form-ctx user (:election electorate))
```
Comment (dzcode) [inline src/voter_registration_works/reg_form.clj:138]: Because of line 136, I realized we need to handle both the `:wy` **and** the `:wy-reg` case here - I'm not sure the best way to structure it, though, so feel free to take a stab or we can pair

File: src/voter_registration_works/reg_form/wy.clj (hunk 7 above)
Comment (mrichards42) [reply src/voter_registration_works/reg_form/wy.clj]: gotcha, yeah I follow

File: src/voter_registration_works/reg_form.clj (hunk 8 above)
Comment (mrichards42) [reply src/voter_registration_works/reg_form.clj:138]: @dzcode I probably don't have all the context for this, so feel free to ignore me if the two of you already have this figured out, but . . . can you explain this? Why are we changing the keyword from `:wy` to `:wy-reg`?

Comment (dzcode) [reply src/voter_registration_works/reg_form.clj]: Yeah - we wound up changing that with Tara's commit [here](https://github.com/democracyworks/voter-registration-works/pull/82/commits/46c5ecbf87c24b7addc7ce29d6452b7a96a79f29). The problem was that we use `:wy-reg` as an alternative to `:nvrf` in `:acceptable-forms` when there's an election, but otherwise key on the state (`:wy`) to determine the form, so need to handle both cases

Comment (dzcode) [reply src/voter_registration_works/reg_form.clj]: If you're curious, this was a long-standing bug explained [here](https://democracyworks.slack.com/archives/C03RNKA6H/p1586362078003300?thread_ts=1586295250.001300&cid=C03RNKA6H)

Comment (mrichards42) [reply src/voter_registration_works/reg_form.clj]: yikes

File: src/voter_registration_works/reg_form.clj [hunk 9]
```diff
@@ -139,9 +135,11 @@
...
                           (:district-division electorate))
                          (user-registration-state user))]
-    (condp = form-type
-      :wy (wy/user->reg-form-ctx user)
-      (nvrf/user-and-election->reg-form-ctx user (:election electorate)))))
+    (cond
+      (#{:wy-reg :wy} form-type) (wy/user-and-election->reg-form-ctx user (:election electorate))
+      (= :nvrf form-type) (nvrf/user-and-election->reg-form-ctx user (:election electorate))
+      :else (do (log/warn (str "unknown reg form type" form-type ". Generating nvrf")) 
+                (nvrf/user-and-election->reg-form-ctx user (:election electorate))))))
```
Comment (mrichards42) [inline src/voter_registration_works/reg_form.clj:142]: This is great and definitely works. It's also a textbook example of `case`:

```clj
(case form-type
  (:wy :wy-reg) (wy/user-and-election->reg-form-ctx user (:election electorate))
  :nvrf (nvrf/user-and-election->reg-form-ctx user (:election electorate))
  ;; no :else keyword needed, but sometimes I'll add a comment:
  ;; else
  (nvrf/user-and-election->reg-form-ctx user (:election electorate)))
```

Comment (mrichards42) [reply src/voter_registration_works/reg_form.clj:142]: Also no `str` needed in logging functions:

```clj
(log/warn "unknown reg form type" form-type ". Generating nvrf")

;; though we're starting to use the `f` versions of logging functions more,
;; which use print-f style args:
(log/warnf "unknown reg form type %s. Generating nvrf" form-type)
```

File: src/voter_registration_works/reg_form.clj [hunk 10]
```diff
@@ -135,13 +131,18 @@
...
+      (case form-type
+        :wy-reg (wy/user-and-election->reg-form-ctx (:election electorate))
+        :nvrg (nvrf/user-and-election->reg-form-ctx user (:election electorate))
+        (do
+          (log/warnf "unknown reg form type %s. Generating nvrf" form-type)
+          (nvrf/user-and-election->reg-form-ctx user (:election electorate)))))
+    (when-let [state (user-registration-state user)]
+      (case state
+        :wy (wy/user-and-election->reg-form-ctx user (:election electorate))
+        (nvrf/user-and-election->reg-form-ctx user (:election electorate))))))
```
Comment (dzcode) [inline src/voter_registration_works/reg_form.clj:145]: One last thing: Let's just use `nil` instead of `(:election electorate)` on 144 + 145, since we know electorate is `nil` anyway

File: test/voter_registration_works/test_utils.clj [hunk 11]
```diff
@@ -3,3 +3,6 @@
 
 (defn lowercase? [string]
   (= string (string/lower-case string)))
+
+(defn deep-merge [m1 m2]
+  (merge-with merge m1 m2))
```
Comment (mrichards42) [inline test/voter_registration_works/test_utils.clj:8]: Nit: this isn't actually deep-merge, since it isn't recursive -- maybe pick a different name (or make it an actual deep merge) to avoid later confusion?

(or pull in medley to get its [deep-merge](https://github.com/weavejester/medley/blob/master/src/medley/core.cljc#L213) without having to write code).

File: src/voter_registration_works/reg_form/wy.clj [hunk 12]
```diff
@@ -10,20 +10,57 @@
...
+     :citizen_no (when (and (not (string/blank? citizen-country))
+                           (not= "US" citizen-country)) "On")}))
+
+(defn dob-and-election-date->age-ctx [dob election-date]
+  {:age_yes (when (form-utils/voter-of-age dob election-date) "On")
+   :age_no (when (form-utils/voter-not-of-age dob election-date) "On")})
+
+(defn phone->ctx [phone]
+  (when phone
+    (when-let [phone (some->> (string/replace phone #"[^0-9]" "")
```
Comment (mrichards42) [inline src/voter_registration_works/reg_form/wy.clj:30]: Where are these unformatted phone numbers coming from? If we need to handle them I think we'll be a lot happier in the long run if we do it the same way [user-works does](https://github.com/democracyworks/user-works/blob/master/src/user_works/util/phone_number.clj), since there are tons of weird phone number edge cases.

Comment (dzcode) [reply src/voter_registration_works/reg_form/wy.clj:30]: They're coming unparsed from turbovote-web, it seems

Comment (mrichards42) [reply src/voter_registration_works/reg_form/wy.clj:30]: fwiw if this is something we just need to get out the door, I'm OK w/ this since I don't think it _breaks_ anything, more that the libphonenumber solution might allow us to catch additional phone numbers that this series of regexes wouldn't. But since the alternative is that we just don't fill in this field, that isn't the end of the world.

Comment (mrichards42) [reply src/voter_registration_works/reg_form/wy.clj:30]: But one thing I just noticed is that we probably want either `re-matches` (or `re-find` with `^$` anchors) so we aren't just taking the first 10 numbers in the string. If the user sent us a bunch of gibberish, we probably don't want to fill this field.

Comment (dzcode) [reply src/voter_registration_works/reg_form/wy.clj:30]: Yeah, since this is an optional field anyway, I'm inclined to leave it, and I'll add a comment in the code that we should circle back to make it more robust if it becomes more critical
